src/cephadm/cephadm

```diff
@@ -3089,6 +3089,80 @@ def command_adopt():
         raise Error('daemon type %s not recognized' % daemon_type)
 
 
+def _adopt_check_online_osd(osd_data_dir, osd_id):
+    # type: (str, str) -> Tuple[Optional[str], Optional[str]]
+
+    osd_fsid, osd_type = None, None
+
+    path = os.path.join(osd_data_dir, 'fsid')
+    try:
+        with open(path, 'r') as f:
+            osd_fsid = f.read().strip()
+        logger.info("Found online OSD at %s" % path)
+        if os.path.exists(os.path.join(osd_data_dir, 'type')):
+            with open(os.path.join(osd_data_dir, 'type')) as f:
+                osd_type = f.read().strip()
+        else:
+            logger.info('"type" file missing for OSD data dir')
+    except IOError:
+        logger.info('Unable to read OSD fsid from %s' % path)
+
+    return osd_fsid, osd_type
+
+
+def _adopt_check_offline_lvm_osd(osd_data_dir, osd_id):
+    # type: (str, str) -> Tuple[Optional[str], Optional[str]]
+
+    osd_fsid, osd_type = None, None
+
+    c = CephContainer(
+        image=args.image,
+        entrypoint='/usr/sbin/ceph-volume',
+        args=['lvm', 'list', '--format=json'],
+        privileged=True
+    )
+    out, err, code = call_throws(c.run_cmd(), verbose=False)
+    if not code:
+        try:
+            js = json.loads(out)
+            if osd_id in js:
+                logger.info("Found offline LVM OSD {}".format(osd_id))
+                osd_fsid = js[osd_id][0]['tags']['ceph.osd_fsid']
+                for device in js[osd_id]:
+                    if device['tags']['ceph.type'] == 'block':
+                        osd_type = 'bluestore'
+                        break
+                    if device['tags']['ceph.type'] == 'data':
+                        osd_type = 'filestore'
+                        break
+        except ValueError as e:
+            logger.info("Invalid JSON in ceph-volume lvm list: {}".format(e))
+
+    return osd_fsid, osd_type
+
+
+def _adopt_check_offline_simple_osd(osd_data_dir, osd_id):
+    # type: (str, str) -> Tuple[Optional[str], Optional[str]]
+
+    osd_fsid, osd_type = None, None
+
+    osd_file = glob("/etc/ceph/osd/{}-[a-f0-9-]*.json".format(osd_id))
+    if len(osd_file) == 1:
+        with open(osd_file[0], 'r') as f:
+            try:
+                js = json.loads(f.read())
+                logger.info("Found offline simple OSD {}".format(osd_id))
+                osd_fsid = js["fsid"]
+                osd_type = js["type"]
+                if osd_type != "filestore":
+                    # need this to be mounted for the adopt to work, as it
+                    # needs to move files from this directory
+                    call_throws(['mount', js["data"]["path"], osd_data_dir])
+            except ValueError as e:
+                logger.info("Invalid JSON in {}: {}".format(osd_file, e))
+
+    return osd_fsid, osd_type
+
```

Comment:
Introducing functions is a great step. You want to go one step further and introduce a class?

```python
class AdoptOsd(object):
    def __init__(self, data_dir_src, daemon_id): ...
    def check_offline_simple_osd(self): ...
```

jsut for the sake of reducing the scope of symbols?

Reply:
Yep, can do. I've also since thought of a couple of other tweaks to make, stay tuned...